src/ethereum/crypto/finite_field.py

```diff
@@ -0,0 +1,399 @@
+"""
+Finite Fields
+^^^^^^^^^^^^^
+"""
+
+# flake8: noqa: D102, D105
+
+import math
+from typing import Iterable, List, Tuple, Type, TypeVar, cast
+
+from typing_extensions import Protocol
+
+from ..base_types import Bytes, Bytes32, Uint
+
+F = TypeVar("F", bound="Field")
+
+
+class Field(Protocol):
+    """
+    A type protocol for defining fields.
+    """
+
+    __slots__ = ()
+
+    @classmethod
+    def zero(cls: Type[F]) -> F:
+        ...
+
+    @classmethod
+    def from_int(cls: Type[F], n: int) -> F:
+        ...
+
+    def __radd__(self: F, left: F) -> F:
+        ...
+
+    def __add__(self: F, right: F) -> F:
+        ...
+
+    def __iadd__(self: F, right: F) -> F:
+        ...
+
+    def __sub__(self: F, right: F) -> F:
+        ...
+
+    def __rsub__(self: F, left: F) -> F:
+        ...
+
+    def __mul__(self: F, right: F) -> F:
+        ...
+
+    def __rmul__(self: F, left: F) -> F:
+        ...
+
+    def __imul__(self: F, right: F) -> F:
+        ...
+
+    def __pow__(self: F, exponent: int) -> F:
+        ...
+
+    def __ipow__(self: F, right: int) -> F:
+        ...
+
+    def __neg__(self: F) -> F:
+        ...
+
+    def __truediv__(self: F, right: F) -> F:
+        ...
+
+
+T = TypeVar("T", bound="PrimeField")
+
+
+class PrimeField(int, Field):
+    """
+    Superclass for integers modulo a prime. Not intended to be used
+    directly, but rather to be subclassed.
+    """
+
+    __slots__ = ()
+    PRIME: int
+
+    @classmethod
+    def from_be_bytes(cls: Type, buffer: "Bytes") -> "Uint":
+        """
+        Converts a sequence of bytes into a element of the field.
+        Parameters
+        ----------
+        buffer :
+            Bytes to decode.
+        Returns
+        -------
+        self : `Uint`
+            Unsigned integer decoded from `buffer`.
+        """
+        return cls(int.from_bytes(buffer, "big"))
+
+    @classmethod
+    def zero(cls: Type[T]) -> T:
+        return cls.__new__(cls, 0)
+
+    @classmethod
+    def from_int(cls: Type[T], n: int) -> T:
+        return cls(n)
+
+    def __new__(cls: Type[T], value: int) -> T:
+        return int.__new__(cls, value % cls.PRIME)
+
+    def __radd__(self: T, left: T) -> T:  # type: ignore[override]
+        return self.__add__(left)
```

Comment:
The issue is that `PrimeField` is subclass of `int` and `GaloisField` is a subclass of `tuple`. This is done as an optimization to prevent another layer of pointer indirection and a malloc. Mypy objects that `__mul__()` is overidden with a method that isn't compatible with the superclass version.

```
src/ethereum/crypto/finite_field.py:108: error: Argument 1 of "__radd__" is incompatible with supertype "int"; supertype defines the argument type as "int"
src/ethereum/crypto/finite_field.py:108: note: This violates the Liskov substitution principle
src/ethereum/crypto/finite_field.py:108: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
```

`PrimeField` and `GaloisField` should be newtypes (like a one element tuple struct in Rust), but Python doesn't have newtypes.